doc/decisions/creating_keys.md

```diff
@@ -0,0 +1,145 @@
+# Creating Keys
+
+## Problem
+
+The API that `libelektra-core` defines for creating Keys must be possible to implement in other languages without complicated workarounds.
+In particular, another implementation should not need to resort to adding a small amount of C to "bridge the gap".
+
+## Constraints
+
+- There should be a way to create a `Key *` with name, value and metadata in a single expression.
+
+  This is needed to allow creating a full `KeySet *` with a single expression, which in turn is very useful e.g. when initializing default configurations.
+
+- Using the escaped name in the constructor function should be avoided, since the unescaping process can be expensive.
+- Ideally, creating a `Key *` without metadata should use at most 3 `malloc`s and 2 `memcpy` to:
+
+  1. `malloc` to allocate the `struct _Key`.
+  2. `malloc` to allocate the key name and `memcpy` to initialize it.
+  3. `malloc` to allocate the key value and `memcpy` to initialize it.
+     This constraint is not absolute, but it should be a goal as long as it doesn't complicate the API to much.
+
+## Assumptions
+
+- Many languages have trouble with C's variadic arguments.
+- Languages will provide a secondary library with more idiomatic APIs on top of `libelektra-core`.
+
+  If necessary, this library will also use unstable APIs from `libelektra-core` and in particular access `struct _Key` directly.
+
+## Considered Alternatives
+
+- Create a constructor function based on heavy use of arrays and `NULL` terminators
+- A simple function that takes `char * name`, `void * value` and `KeySet * meta`.
+
+  The size of the name and value would have to be passed separately (a potential source of errors).
+
+## Decision
+
+`libelektra-core` only provides the absolute minimal API to create a `Key *`:
+
+```c
+/**
+ * Allocates a new key with the name "/" and returns it.
+ *
+ * The key will have no value and no metadata.
+ */
+Key * keyCreate (void);
```

Comment:
> this is what users should use. 

You mean C users? If everyone should use the other form, it should be in the core.

Anyway, I don't see a reason for having a keyNew without arguments because we decided that every key must have a name as invariant. So the minimal form is keyNew (const char* key);

Reply:
> You mean C users? If everyone should use the other form, it should be in the core.

Yes, `libelektra-operations.so` (and therefore `keyNew`/`ksNew`) is just for C users. Other implementations and bindings should provide other APIs for suitable for their language on top of `libelektra-core.so` (i.e. on top of `keyCreate`/`ksCreate`).

> Anyway, I don't see a reason for having a keyNew without arguments because we decided that every key must have a name as invariant.

It still has the name `"/"` (cascading root key) which kind of works like a "null" value for key names. It is the simplest possible name and could be optimized to not use `malloc`.

> So the minimal form is keyNew (const char* key);

In C we need the size (see also https://github.com/ElektraInitiative/libelektra/issues/3529#issuecomment-1023616182). Adding a size argument that can result in errors and is just annoying for other languages doesn't seem like a good idea for the core API. Therefore, I decided the `keyCreate (void)` API is the best solution.